internal/process/process.go

```diff
@@ -78,8 +78,6 @@ func (p *Process) Stop() {
 		// signal our intent to shutdown and not restart the process
 		p.cancel()
 	}
-	// close stdin so the app can shut down gracefully.
-	p.Stdin.Close()
```

Comment:
It's important to notify the process that it needs to stop. You don't want the first notification to the process that it needs to stop to be trying to terminate the process, otherwise you drop metrics and don't allow it to finish up the work it's doing. Instead you need to close the pipe, give it time to process what's in process, and then close gracefully before being sent termination signals. We can't remove this.

Reply:
Ok, fair point. Nonetheless, this way the error in the test: `Error reading stdout: read |0: file already closed` will not go away.

Reply:
error reading **stdout** though, sounds like something else, not this stdin pipe.

Reply:
Yes, but the error does go away when the `p.Stdin.Close()` is removed.

I've reverted the change now but I'd say that this might require a closer look.

For future reference: the error is coming from: https://github.com/influxdata/telegraf/blob/de313fcde6c59376208a07ca9ace88aecaaf42a7/plugins/processors/execd/execd.go#L134

which is basically reading what `runCountMultiplierProgram()` prints out
https://github.com/influxdata/telegraf/blob/da1cb18df788a5b7a522c34b41e13c1264c8672b/plugins/processors/execd/execd_test.go#L104-L145

and when the stdin is being closed I get an error in https://github.com/influxdata/telegraf/blob/da1cb18df788a5b7a522c34b41e13c1264c8672b/plugins/processors/execd/execd_test.go#L110-L113

Reply:
it goes away because you're no longer shutting down the process after reading the first metric, but this is what I'm intending and testing. This should work: 
- write 10 metrics
- read 1
- close() _(stop the plugin)_
- read the remaining 9 metrics

I believe the real issue is that in some cases the `cmdReadOut` goroutine+func isn't running until after the process is already shutting down, so the read runs into the error. Ideally I'd like to wait until that goroutine starts before triggering shutdown.

Ignoring this specific error is an option but then we're potentially missing stdout/stderr output.

What OS are you using? I'm assuming you're able to reliably reproduce the failure?

Reply:
I use the following one liner which reliably gives me an error (with a passing test) every 2-3 seconds or so.

I'm running this on mac ATM.

Reply:
Ok I did some debugging and I was seeing this error come up at the end of the stream after the data was already read. If that's the case, it's not a real error at all, it's just indicating that the pipe is closed, and can be safely ignored

Reply:
this is good. If the pipe was being closed before the data was read, we'd actually see the test fail, so we can be pretty confident it's doing the right thing.

Reply:
Okay... so I'm still seeing cases where some of the metrics are missing afterwards. Maybe the pipe's being closed with metrics still in the pipe? I'm missing something.

Reply:
> I'm still seeing cases where some of the metrics are missing afterwards
I'll try to take a look at this tomorrow. Can you share your approach of testing this?

Reply:
I'm using `go test github.com/influxdata/telegraf/plugins/processors/execd -count=100 -v -timeout 60s`, and adjusting count for about 15s of runtime.

Here's the diff of my workspace with local changes. It's not quite right as it's still failing periodically, though maybe not as much. https://gist.github.com/ssoroka/fe9c604f6b3cee6dc965948d43c20ee5

I'm not a fan of the sleep, I'm just testing with it. There might be something going on where the process ends before stdout has been completely read and it tears down the pipes or something. Not sure. I'm under the impression that this should work without changing the basic structure of the test (and really it does like 99.9% of the time).

Reply:
(I also noticed that `acc.Wait(9)` should be `acc.Wait(10)` as it's the total count, not the count since last call.

Reply:
After look at it again I do believe that this not a problem (the error returned at https://github.com/influxdata/telegraf/blob/de313fcde6c59376208a07ca9ace88aecaaf42a7/plugins/processors/execd/execd.go#L134)

When `p.Stdin.Close()` is called the loop in the test https://github.com/influxdata/telegraf/blob/da1cb18df788a5b7a522c34b41e13c1264c8672b/plugins/processors/execd/execd_test.go#L109-L113 is terminated via a `return` and my suspicion is that sometimes this will be faster than the stderr/stdout read in `cmdReadErr`/`cmdReadOut` where it will return `file already closed`.

I'd say it is safe to ignore this specific error. As a safeguard one could add a check whether the process has already returned and if so then don't report `os.ErrClosed` like so:

```
func (e *Execd) cmdReadOut(out io.Reader) {
	scanner := bufio.NewScanner(out)

	for scanner.Scan() {
		metrics, err := e.parser.Parse(scanner.Bytes())
		if err != nil {
			e.Log.Errorf("Parse error: %s", err)
		}

		for _, metric := range metrics {
			e.acc.AddMetric(metric)
		}
	}

	if err := scanner.Err(); err != nil {
		ps := e.process.Cmd.ProcessState
		// If the process already returned don't report ErrClosed errors
		if (ps != nil && ps.Exited()) && (errors.Unwrap(err) == os.ErrClosed) {
			return
		}

		e.Log.Errorf("Error reading stdout: %s", err)
	}
}
```